### PR TITLE
[SMP-998] Api Compliance

### PR DESCRIPTION
**Pull request by ppoursafa**
https://dl.polyconseil.fr/jira/browse/SMP-998

### CODE REVIEW

File: mds/authent/migrations/0003_auto_20190903_1529.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        ('authent', '0002_auto_20190124_0928'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='application',
+            name='owner',
+            field=models.UUIDField(blank=True, db_index=True, help_text='Unique identifier for the owner of the application', null=True),
+        ),
+    ]
```
Comment (bors-ltd) [inline mds/authent/migrations/0003_auto_20190903_1529.py]: You got this migration because you typed `./manage.py migrate` with no argument, which is bad practice.

Always type the app you want to migrate as the first argument. You shouldn't have any migration in authent, because you didn't alter it.

File: mds/factories.py [hunk 2]
```diff
@@ -748,3 +748,18 @@ def prev_policies(self, create, extracted, **kwargs):
...
             for policy in extracted:
                 self.prev_policies.add(policy)
+
+
+class ComplianceFactory(factory.django.DjangoModelFactory):
+
+    policy = factory.SubFactory(Policy)
+    vehicle = factory.SubFactory(Device)
+    rule = factory.Faker("uuid4")
+    geography = factory.Faker("uuid4")
```
Comment (bors-ltd) [inline mds/factories.py]: :+1:

File: mds/migrations/0044_auto_20190906_0818.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Generated by Django 2.1.11 on 2019-09-06 08:18
```
Comment (bors-ltd) [inline mds/migrations/0044_auto_20190906_0818.py]: Rename the automatically generated filename with an explicit name.

File: mds/models.py [hunk 4]
```diff
@@ -329,3 +329,20 @@ class Policy(models.Model):
...
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4)
+    policy = models.ForeignKey(
+        Policy, related_name="compliances", on_delete=models.CASCADE
+    )
+    vehicle = models.ForeignKey(
+        Device, related_name="compliances", on_delete=models.CASCADE
+    )
+    rule = models.UUIDField(default=uuid.uuid4)
+    geography = models.UUIDField(default=uuid.uuid4)
+    start_date = models.DateTimeField(blank=False, null=False)
```
Comment (bors-ltd) [inline mds/models.py]: Don't say a field is not null, this is the default behaviour.

File: mds/apis/agency_api/v0_x/compliance.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+from mds import models
+
+from datetime import datetime
+import pytz
+
+
+class ComplianceSerializer(serializers.ModelSerializer):
+    rules_id = serializers.SerializerMethodField()
+
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: You can remove the empty line here

File: mds/apis/agency_api/v0_x/compliance.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    compliances = serializers.SerializerMethodField()
+
+    class Meta:
+        model = models.Policy
+        fields = ("id", "rules_id", "compliances")
+
+    def get_rules_id(self, policy):
+        rules = []
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: There is a way to write the same this in one line :  
`return [rule["rule_id"]] for rule in policy.rules]`

This is what we call a one-liner

File: mds/apis/agency_api/v0_x/compliance.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        for rule in policy.rules:
+            rules.append(rule["rule_id"])
+        return rules
+
+    def get_compliances(self, policy):
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: I can't see where this is called from. 
Maybe some code that you don't need anymore ?

File: mds/apis/agency_api/v0_x/compliance.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        if end_date:
+            end_date = datetime.fromtimestamp(int(end_date), tz=pytz.UTC)
+        if provider_id and end_date:
+            queryset = queryset.filter(
+                Q(compliances__end_date__gte=end_date)
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: I think you can use `.exclude(compliances__end_date__lt=end_date)`

File: mds/apis/agency_api/v0_x/compliance.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                    "compliances",
+                    queryset=models.Compliance.objects.filter(
+                        vehicle__provider__id=provider_id
+                    ),
+                    to_attr="compliances_pref",
+                )
+            )
+        elif end_date:
+            queryset = queryset.filter(
+                Q(compliances__end_date__gte=end_date)
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: Same here

File: mds/apis/agency_api/v0_x/compliance.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                | Q(compliances__end_date__isnull=True),
+                compliances__vehicle__provider__id=provider_id,
+                compliances__start_date__lte=end_date,
+            ).prefetch_related(
+                Prefetch(
+                    "compliances",
+                    queryset=models.Compliance.objects.filter(
+                        Q(end_date__gte=end_date) | Q(end_date__isnull=True),
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: same here

File: mds/apis/agency_api/v0_x/urls.py [hunk 11]
```diff
@@ -11,11 +11,16 @@
 from . import policies
 from . import service_areas
 from . import vehicles
+from . import compliance
 
 
 agency_router = routers.DefaultRouter()
 agency_router.register(r"vehicles", vehicles.DeviceViewSet, basename="device")
 agency_router.register(r"policies", policies.PolicyViewSet, basename="policy")
+agency_router.register(
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/urls.py]: Depending on if we stay under 80 or 120 chars per line, you may keep this on one line.
I don't know what coding rule we use.

File: mds/authent/migrations/0003_auto_20190903_1529.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.1.11 on 2019-09-03 15:29
```
Comment (tonial) [inline mds/authent/migrations/0003_auto_20190903_1529.py]: You should remove this migration from your code : you didn't change `mds/authent/models.py` file.
As @bors-ltd said, always specify the app name when using `./manage.py migrate` to prevent django from creating migrations on the project's dependencies

File: mds/migrations/0044_compliance.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# Generated by Django 2.1.11 on 2019-09-09 14:34
```
Comment (tonial) [inline mds/migrations/0044_compliance.py]: You should give a even clearer name : 
0044_add_compliance_model.py so that next time someone edits compliance model, he would be able to call the migration `xxxx_add_compliance_<field_name>.py` or so

File: mds/migrations/0044_compliance.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            name='Compliance',
+            fields=[
+                ('id', models.UUIDField(default=uuid.uuid4, primary_key=True, serialize=False)),
+                ('rule', models.UUIDField(default=uuid.uuid4)),
+                ('geography', models.UUIDField(default=uuid.uuid4)),
+                ('start_date', models.DateTimeField()),
+                ('end_date', models.DateTimeField(blank=True, null=True)),
+            ],
+        ),
+        migrations.AlterField(
```
Comment (tonial) [inline mds/migrations/0044_compliance.py]: You may clean the migration if you want : the following AlterField are not mandatory (since not added in models.py) 
It is no mandatory though, I tend to do it because my commit will be cleaner :)

File: tests/apis/agency_api/v0_x/test_compliance.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        vehicle_id=factories.Device().id,
+        start_date=datetime.datetime(2009, 12, 6, 16, 29, 43, 79043, tzinfo=pytz.UTC),
+        end_date=datetime.datetime(2009, 12, 7, 16, 29, 43, 79043, tzinfo=pytz.UTC),
+    )
+
+    # Test without auth
+    n = BASE_NUM_QUERIES
+    n += 1  # query on device
+    n += 1  # query on policy  # query on Compliance
+    n += 1  # query on second compliance
```
Comment (tonial) [inline tests/apis/agency_api/v0_x/test_compliance.py]: :+1:

File: mds/apis/agency_api/v0_x/compliance.py (hunk 6 above)
Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: Propre

File: mds/apis/agency_api/v0_x/compliance.py (hunk 5 above)
Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: C'est fait

File: mds/models.py (hunk 4 above)
Comment (ppoursafa) [reply mds/models.py]: C'est réglé

File: mds/migrations/0044_auto_20190906_0818.py (hunk 3 above)
Comment (ppoursafa) [reply mds/migrations/0044_auto_20190906_0818.py]: C'est fait ;)

File: mds/factories.py (hunk 2 above)
Comment (ppoursafa) [reply mds/factories.py]: merci :)

File: mds/authent/migrations/0003_auto_20190903_1529.py (hunk 1 above)
Comment (ppoursafa) [reply mds/authent/migrations/0003_auto_20190903_1529.py]: Merci pour l'ifno

File: tests/apis/agency_api/v0_x/test_compliance.py (hunk 15 above)
Comment (ppoursafa) [reply tests/apis/agency_api/v0_x/test_compliance.py]: Merci :)

File: mds/apis/agency_api/v0_x/compliance.py (hunk 8 above)
Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: Mieux vaut utiliser exclude que filter ?

File: mds/apis/agency_api/v0_x/compliance.py [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        if end_date:
+            end_date = datetime.fromtimestamp(int(end_date), tz=pytz.UTC)
+        if provider_id and end_date:
+            queryset = (
+                queryset.exclude(compliances__end_date__lt=end_date)
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: For a better reading experience, I would write :
```
queryset = (
    queryset
    .exclude(...)
    .filter(
        ...
    )
    .prefetch_rekated(
```

File: mds/apis/agency_api/v0_x/compliance.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+                            end_date__lte=end_date
+                        ).filter(
+                            vehicle__provider__id=provider_id, start_date__lt=end_date
+                        ),
+                        to_attr="compliances_pref",
+                    )
+                )
+            )
+        elif provider_id:
+            queryset = queryset.filter(
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: same comment here

File: mds/apis/agency_api/v0_x/compliance.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+                            match["measured"] += 1
+                            break
+                    else:
+                        current_compliance["matches"].append(
+                            {"geography": str(compliance.geography), "measured": 1}
+                        )
+                    current_compliance["vehicles_in_violation"].append(
+                        str(compliance.vehicle.id)
+                    )
+                    current_compliance["total_violations"] = len(
```
Comment (Ja-Bert) [inline mds/apis/agency_api/v0_x/compliance.py]: You may compute every "total_violations" at the end of your function
This would avoid getting through every vehicles in violations many many times

File: mds/apis/agency_api/v0_x/compliance.py (hunk 7 above)
Comment (tonial) [reply mds/apis/agency_api/v0_x/compliance.py]: Oh, compliance isn't a real attribute of Policy... :+1:

File: mds/apis/agency_api/v0_x/compliance.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+class ComplianceViewSet(viewsets.ModelViewSet):
+    queryset = models.Policy.objects
+    serializer_class = ComplianceSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        if end_date:
+            end_date = datetime.fromtimestamp(int(end_date), tz=pytz.UTC)
```
Comment (Ja-Bert) [inline mds/apis/agency_api/v0_x/compliance.py]: .exclude(...date_lt=...) is the same as .filter(...date_gte=...) 

So you could make a dictionnary in your ifs and pass **filter_dict to your filter

```
filter_dict = {"compliances__end_date__gte": end_date, "compliances__vehicle__provider__id": provider_id...}
queryset.filter(**filter_dict)
```

File: mds/apis/agency_api/v0_x/compliance.py (hunk 16 above)
Comment (tonial) [reply mds/apis/agency_api/v0_x/compliance.py]: You can let it like that, if it's what black chooses

File: mds/apis/agency_api/v0_x/compliance.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+        else:
+            filter_prefetch = Prefetch(
+                "compliances",
+                queryset=models.Compliance.objects.filter(),
+                to_attr="compliances_pref",
+            )
+
+        if (provider_id and end_date) or end_date:
+            queryset = (
+                queryset.exclude(compliances__end_date__lt=end_date)
```
Comment (Ja-Bert) [inline mds/apis/agency_api/v0_x/compliance.py]: en mettant un "...gte" dans tes filters tu peux retirer ce exclude et donc faire
```suggestion
if provider_id or end_date:
    queryset = queryset.filter(**filters).prefetch_related(filter_prefech)
else:
    queryset = queryset.prefetch_related(filter_prefech)
```

Et vu que tu as tes filters initialisés à un array vides tu pourras même faire directement

```suggestion
queryset = queryset.filter(**filters).prefetch_related(filter_prefech)
```

File: mds/apis/agency_api/v0_x/compliance.py (hunk 19 above)
Comment (tonial) [reply mds/apis/agency_api/v0_x/compliance.py]: nope :)
because end_date may be null
=> @ppoursafa you should maybe add or adapt test to be sure that the case with a object with end_date=None is handled correctly, to prevent regressions in the future

File: mds/apis/agency_api/v0_x/compliance.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    serializer_class = ComplianceSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        filters = {}
+        if provider_id and end_date:
+            filters["compliances__vehicle__provider__id"]: provider_id
+            filters["compliances__start_date__lte"]: end_date
```
Comment (tonial) [inline mds/apis/agency_api/v0_x/compliance.py]: didn't you forget the compliances with end_date=None ?

Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: Thanks to the .exclude it delete all the end_date less that the end_date put in the call API. So he keep all the end_date higher and None. I make a little bit time to understand that yesterday but it's tested and approved :)

Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: If you want i can add that in the test_compliance

File: mds/apis/agency_api/v0_x/compliance.py (hunk 19 above)
Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: None is handled correctly so i let exclude ;)

File: mds/apis/agency_api/v0_x/compliance.py (hunk 21 above)
Comment (ppoursafa) [reply mds/apis/agency_api/v0_x/compliance.py]: i put that in test_compliance ;)
